### PR TITLE
Collapse siblings when expanding one item

### PR DESCRIPTION
**Pull request by fanwgwg**
[*DYN - 823 Implement expanding of one item collapses the other siblings*](https://jira.autodesk.com/browse/DYN-823)

capture2

### Reviewer 
@Benglin 

### FYI
@riteshchandawar

### CODE REVIEW

File: src/components/LibraryItem.tsx [hunk 1]
```diff
@@ -258,6 +263,14 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
         }
     }
 
+    // Unexpand all other slibling items when one item is being expanded
```
Comment (Benglin) [inline src/components/LibraryItem.tsx]: Should be *collapse all other siblings when an item is being expanded*

Comment (Benglin) [inline src/components/LibraryItem.tsx]: Wait, I think this does something entirely different. Let's rephrase it this way:

```js
// Collapse all child items when one of the child items is expanded
onSingleChildItemExpanded() {
  for (...) { }
  this.setState({ expanded: true }); // Make the current item (parent) expanded.
}
```

File: src/components/LibraryItem.tsx [hunk 2]
```diff
@@ -249,6 +251,9 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
     onLibraryItemClicked() {
         // Toggle expansion state.
         let currentlyExpanded = this.state.expanded;
+        if (this.props.data.childItems.length > 0 && !currentlyExpanded) {
+            this.props.onExpand();
+        }
```
Comment (Benglin) [inline src/components/LibraryItem.tsx]: The notification is sent before expansion actually take place, so let's rename this from `onExpand` to `onItemWillExpand` (just to borrow from React.js term `componentWillReceiveProps` and `componentWillUnmount`, etc).

File: src/components/LibraryItem.tsx (hunk 1 above)
Comment (Benglin) [inline src/components/LibraryItem.tsx]: Wait, it should be named `onSingleChildItemWillExpand`, and not `onSingleChildItemExpanded`.

**Comment by Benglin** [comment]
Thanks for this, @fanwgwg! 👍 minor name changes and we'll go with it.

**Comment by fanwgwg** [comment]
@Benglin I've made the changes accordingly! Thanks for the review!

File: src/components/LibraryItem.tsx [hunk 3]
```diff
@@ -193,7 +194,12 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
                     // 'getNestedElements' method is meant to render all other 
                     // types of items except ones of type create/action/query.
                     regularItems.map((item: ItemData) => {
-                        return (<LibraryItem key={index++} libraryContainer={this.props.libraryContainer} data={item} />);
+                        return (<LibraryItem
+                            key={index++}
+                            libraryContainer={this.props.libraryContainer}
+                            data={item}
+                            onItemWillExpand={this.onItemWillExpand.bind(this)}
+                        />);
```
Comment (Benglin) [inline src/components/LibraryItem.tsx]: Lining up this way is much easier to read! 👍

File: src/components/LibraryItem.tsx [hunk 4]
```diff
@@ -258,6 +268,14 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
         }
     }
 
+    // Collapse all child items when one of the child items is expanded
+    onItemWillExpand() {
```
Comment (Benglin) [inline src/components/LibraryItem.tsx]: Can we rename as I originally recommended? This is at the parent level and we need to make that clear.

**Comment by fanwgwg** [comment]
@Benglin  Ok I've just renamed the function

**Comment by mjkkirschner** [comment]
curious about this - is this an option that can be set at construction time or the only behavior?

**Comment by Benglin** [comment]
Hi @mjkkirschner, do you mean `onItemWillExpand` behaviour or the single-expansion behaviour? Feel free to ping me on Slack, I can chat :)

**Pull request merged**